### PR TITLE
Updates with token management

### PR DESCRIPTION
**Pull request by albaintor**
The token and HA endpoint are loaded at startup (unchanged)
Now the token and HA endpoint will also be loaded (if not) or reloaded (if HA connected) if the HA token or endpoint have changed : this can occur in 3 situations :
1/ When the user from HA  registers a token and create a HA driver instance through rest API (initial config flow from Home assistant)
2/ When the user from HA reconfigure the entities from HA component : this will regenerate a new token (even if one exists) : this use case will also be useful if the user accidentally deletes the token from HA, or if the token expires
3/ When the user wants to migrate from one HA instance to another (hardware updgrade, migrate from docker to dedicated hardware...)
To summarize the logic, after each Connect event from the remote :
1/ Driver checks if configuration has changed (token or HA url)
2/ If yes, disconnect & reconnect the HA client with the new setup
3/ Otherwise keep the same logic (try to connect or just notify current state)

### CODE REVIEW

**Comment by albaintor** [comment]
Also some things need to be modified in the core API (not blocking I think) :
- `get_runtime_info` request to the R2 -->  this message type should be replaced with an event instead of a request as the driver acts like a client. I wrote comments in the code for this
- When the `runtime_info` structure is received from the remote, there is not the expected "remote id" we discussed earlier. I only get this info :
```json{"driver_id": "hass", "intg_ids": ["hass.main"], "extra": {} }```

I expected a remote unique identifier to be sent then to the HA component through websocket : this is necessary to be able to make the difference between 2 remotes that send ws messages to home assistant.

**Comment by zehnm** [comment]
> * `get_runtime_info` request to the R2 -->  this message type should be replaced with an event instead of a request as the driver acts like a client. I wrote comments in the code for this. FYI I call this request after the connection from the remote (r2_connection.rs).

Why exactly should it be an event? I couldn't find the related comments.
Both the driver and the Remote can send requests to the other party.
Or do you mean that it should be sent automatically as an event by the Remote after connection?

> * When the `runtime_info` structure is received from the remote, there is not the expected "remote id" we discussed earlier.

The remote identifiers (hostname and MAC address) are not returned in the `get_runtime_info`, but the `get_version` response:

https://github.com/unfoldedcircle/core-api/blob/46dee3be89092324d1e0e54f14add6edcb3e3c04/integration-api/UCR-integration-asyncapi.yaml#L2486-L2501

**Review by zehnm — CHANGES_REQUESTED** [review]
Sorry for the late review, there were no rainy days during my vacations :-)

I don't see a need for the URL & token change check & logic. It makes it more complicated, and most use cases are already handled with the reconnect logic. Whenever a new connection to HA is established, the current configuration is retrieved.
Let me know if I've overseen something.

File: src/controller/handler/r2_connection.rs [hunk 1]
```diff
@@ -3,8 +3,34 @@
...
+            msg: Some(f.msg),
+            msg_data: f.msg_data,
+            req_id: None,
+            code: None,
+            extra: HashMap::new(),
+            ts: None,
+            cat: None,
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (zehnm) [reply src/controller/handler/r2_connection.rs:33]: Added in 0.12.1: https://github.com/unfoldedcircle/api-model-rs/commit/1df137550bbae6131a128624726f048c7ca6f136

The above code block can be removed.

**Comment by albaintor** [comment]
> > * `get_runtime_info` request to the R2 -->  this message type should be replaced with an event instead of a request as the driver acts like a client. I wrote comments in the code for this. FYI I call this request after the connection from the remote (r2_connection.rs).
> 
> Why exactly should it be an event? I couldn't find the related comments. Both the driver and the Remote can send requests to the other party. Or do you mean that it should be sent automatically as an event by the Remote after connection?
> 
> > * When the `runtime_info` structure is received from the remote, there is not the expected "remote id" we discussed earlier.
> 
> The remote identifiers (hostname and MAC address) are not returned in the `get_runtime_info`, but the `get_version` response:
> 
> https://github.com/unfoldedcircle/core-api/blob/46dee3be89092324d1e0e54f14add6edcb3e3c04/integration-api/UCR-integration-asyncapi.yaml#L2486-L2501

Sorry forget this comment : I realized after that the remote id was available in the get-version message. Also I am not a websocket expert, and indeed both client/servers can send requests. However I reported an issue with the UC rust sdk that's missing primitive to send a request (request should derive from message and let generate a request structure with an impl method)

File: src/configuration.rs [hunk 2]
```diff
@@ -187,6 +187,19 @@ impl HomeAssistantSettings {
...
+        let url = self.get_url();
+        if url.as_str() != self.url.as_str() {
+            return true;
+        }
+        let token = self.get_token();
+        if token != self.token {
+            return true;
+        }
+        false
+    }
```
[the remark this answers is not in the document]
Comment (albaintor) [reply src/configuration.rs:201]: Are you sure about that ? ok the env variables change are not check but this is not necessary. Your code reload the HA url file and token file at each call so if the file content is modified, the new URL & token will be caught ?

```rust
fn get_token_value(&self, key: &str) -> Option<String> {
        let mut path = PathBuf::from(env::var(ENV_TOKENS_HOME).ok()?);
        path.push(key);
        if !path.is_file() {
            warn!("Token file '{key}' does not exist. Using local configuration.");
            return None;
        }

        match fs::read_to_string(path) {
            Ok(v) => Some(v.trim().to_string()),
            Err(e) => {
                error!("Error reading token file '{key}', using local configuration. {e}");
                None
            }
        }
    }
```

File: src/client/set_remote_id.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2022 Unfolded Circle ApS, Markus Zehnder <markus.z@unfoldedcircle.com>
+// SPDX-License-Identifier: MPL-2.0
+
+//! Actix actor handler implementation for the `GetStates` message
```
[the remark this answers is not in the document]
Comment (albaintor) [reply src/client/set_remote_id.rs:4]: Ok changed

File: src/controller/handler/r2_event.rs [hunk 4]
```diff
@@ -4,53 +4,126 @@
...
+                        "[{}] Not connected, requesting connection to HA client :  {}",
+                        msg.ws_id, self.device_state
+                    );
+                    ctx.notify(ConnectMsg::default());
+                    self.send_device_state(&msg.ws_id);
+                    return Box::pin(fut::ready(()));
+                } else {
+                    // HA device already connected and configuration unchanged, just notify state
+                    self.send_device_state(&msg.ws_id);
+                    return Box::pin(fut::ready(()));
```
[the remark this answers is not in the document]
Comment (albaintor) [reply src/controller/handler/r2_event.rs:99]: Sorry if I am wrong because it has been already a few weeks I have been working on this :-) : before those modifications I remember that I was struggled with the machine state in a "require setup" state that I was not able to unlock through APIs.

My test (see https://discord.com/channels/@me/1258748339344904222/1281899277999673417) : 
1. HA driver is not configured (no URL/token), registered as an external driver but this does not matter
2. HA component pushes the token and URL to the remote through API and just before I created the 2 files HA token and url pointed by the ENV variable.
3. HA component registers HA driver through API
4. Without this modification : the HA driver remain in "require setup" state and cannot be configured

File: src/client/mod.rs [hunk 5]
```diff
@@ -313,7 +327,7 @@ impl HomeAssistantClient {
                         );
                         self.unsubscribe_uc_events(ctx);
                         self.subscribe_uc_events(ctx);
-                    }
+                    }*/
```
[the remark this answers is not in the document]
Comment (albaintor) [reply src/client/mod.rs:330]: This code caused a problem : if there were 0 subcribed entities, the HA driver didn't subscribe to events for these "0 entities" which is normal because we won't expect any event from 0 entities. But this event is expected by the unified integration to list the subscribed entities, even if there are none. I deleted this commented code

File: src/controller/handler/r2_connection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (albaintor) [reply src/controller/handler/r2_connection.rs:33]: Ok done

File: src/configuration.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zehnm) [reply src/configuration.rs:201]: Yes I'm sure :-)
- `HomeAssistantSettings::url` && `HomeAssistantSettings::token` fields are private and cannot be accessed outside of configuration.rs
  - These fields only hold the local configuration values, not from the ENV variables.
- The `ConnectMsg` handler implementation to connect to HA server in src/controller/handler/ha_connection.rs calls `self.settings.hass.get_url()` and `self.settings.hass.get_token()` methods.
  - These methods check & read the ENV files first, with a fallback to the local configuration.

Migrating or switching to another HA server is not a common use-case done often.
- Migration to another server with decommissioning the old one is handled with the current logic.
- Switching to another HA server requires a reboot or the remote going to sleep & wakeup again to trigger a reconnect. I.e. also "half-way" handled.

If anything shouldn't work as described, then it's a bug and requires fixing.

**Comment by zehnm** [comment]
> However I reported an issue with the UC rust sdk that's missing primitive to send a request

See comment above, just added today in 0.12.1 ;-)

File: src/controller/handler/r2_connection.rs [hunk 6]
```diff
@@ -14,6 +40,19 @@ impl Handler<NewR2Session> for Controller {
...
+        if let Some(session) = self.sessions.get_mut(&msg.id) {
+            let request_id = session.new_msg_id();
+            if let Ok(request) = WsRequest::new(request_id, "get_version", json!({})) {
+                let message = <WsMessage as r2_connection::From<WsRequest>>::from(request);
+                match session.recipient.try_send(SendWsMessage(message)) {
+                    Ok(_) => info!("[{}] Request sent", request_id),
+                    Err(e) => error!("[{}] Error sending entity_states: {e:?}", msg.id),
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (albaintor) [reply src/controller/handler/r2_connection.rs:55]: done !

**Comment by albaintor** [comment]
I just pushed a new commit with the modifications requested

File: src/configuration.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (albaintor) [reply src/configuration.rs:201]: Ok I'll remove this check and test again the whole chain and try to reproduce the issue that I had

File: src/controller/handler/r2_event.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (albaintor) [reply src/controller/handler/r2_event.rs]: Yes exactly, the machine state needs to be updated (following the driver registration from HA component) in order to handle available/subscribed entities messages.
I rollbacked all this part (and removed connection_settings_changed), so now it should not work anymore until you find some way to update the machine state.

Meantime, after some thoughts, I decided to make permanent checks for custom events instead during the first 10 minutes after HA start, because the unified HA component could be installed at any time (so after the 10 minutes) and won't be detected.  To be discussed

**Comment by albaintor** [comment]
New push with rollback of the (re)connect modifications. Something need to be done after machine state : to be updated after registration to allow available/subscribed entities messages

File: src/client/mod.rs [hunk 7]
```diff
@@ -125,6 +131,8 @@ impl HomeAssistantClient {
                 subscribed_entities: HashSet::new(),
                 authenticated: false,
                 remote_id: "".to_string(),
+                uc_ha_component_check_interval: Duration::from_secs(5),
+                uc_ha_component_check_duration: Duration::from_secs(0), // check forever
```
Comment (zehnm) [inline src/client/mod.rs:135]: Magic numbers for special modes should be avoided in Rust, better use `Option<Duration>` instead for an unset duration.

The infinite polling check should be avoided, better stop after a certain duration.
- usually the Remote is only connected for a few minutes before it goes into standby
- connection to HA server is disconnected when the Remote goes into standby
- after wake-up a new polling cycle starts

I prefer to keep it as simple as possible, e.g. give clear setup instructions in the HA component and show the status in the UC HA integration setup.

--> We can leave it for now until everything is working, but we have to clean up the polling mechanism and fix the potential race condition in `check_uc_ha_component`.

File: src/client/mod.rs [hunk 8]
```diff
@@ -634,6 +675,44 @@ impl HomeAssistantClient {
         }
         self.subscribe_uc_events_id = None;
     }
+
+    /// Check after UC HA component regularly
+    fn check_uc_ha_component(
+        &mut self,
+        _ctx: &mut Context<HomeAssistantClient>,
```
Comment (zehnm) [inline src/client/mod.rs:682]: Cosmetic: `ctx` parameter is used: remove underscore prefix. (Underscore prefix is used for unreferenced variables).

File: src/client/mod.rs [hunk 9]
```diff
@@ -634,6 +675,44 @@ impl HomeAssistantClient {
...
+        ha_start_time: SystemTime,
+    ) {
+        if !_ctx.connected() || self.uc_ha_component {
+            debug!("UC HA component found");
+            return;
+        }
```
Comment (zehnm) [inline src/client/mod.rs:688]: Misleading log message if disconnected and uc_ha_component is false. Create two separate checks, just return if not connected, if `uc_ha_component` is true: log & return

File: src/client/mod.rs [hunk 10]
```diff
@@ -634,6 +675,44 @@ impl HomeAssistantClient {
...
+                }
+                Err(e) => {
+                    error!(
+                        "Error during UC HA component check, stop checking after {:?}",
+                        e
+                    );
+                }
+            }
+        }
+        _ctx.run_later(self.uc_ha_component_check_interval, move |act, ctx| {
```
Comment (zehnm) [inline src/client/mod.rs:706]: This can create a race condition with multiple pollers running at the same time:
- recursive call of `check_uc_ha_component`
- `check_uc_ha_component` is also called whenever `auth_ok` message is received

The returned `SpawnHandle` must be stored and cancelled if set.

--> I can fix this after merge. We might have to change more things for the UC component check.

**Review by zehnm — CHANGES_REQUESTED** [review]
Please fix the cosmetic code issues, without the mentioned race condition & duration handling, unless of course you want to :-)

Then I'll merge it into my branch and check the remaining issues and make it work with the UC HA component.

**Review by zehnm — APPROVED** [review]
Thanks a lot!
I will merge it into my branch and start testing today with the UC HA component.

**Comment by albaintor** [comment]
Hi Markus, no problem :)
I will answer to your last comment on discord
